File: MK-52_ESP32/src/Receiver_DATA_K.cpp
//////////////////////////////////////////////////////////
//
//  MK-52 RESURRECT
//  Copyright (c) 2020 Mike Yakimov.  All rights reserved.
//  See main file for the license
//
//////////////////////////////////////////////////////////

#include "Receivers.hpp"
#include "../functions/functions.h"

#define __DEBUG

using namespace MK52_Interpreter;

unsigned long Receiver_DATA_K::init( void *components[]) {
    #ifdef __DEBUG
    Serial.println( "Init DATA_K");
    #endif
    return Receiver::init(components);
}

void Receiver_DATA_K::activate( uint8_t scancode, int8_t parent){
    #ifdef __DEBUG
    Serial.println( "Activating receiver DATA_K");
    #endif
    Receiver::activate( scancode, parent);
    _lcd->updateStatusFMODE( " K ");
    _mode = 1;
    if(!scancode) return;
    tick(scancode);
}

int Receiver_DATA_K::tick( uint8_t scancode){
    int return_value = COMPONENT_RECEIVER_DATA_N;
    switch( scancode){
        // Column 0
        case 1:
            return_value = COMPONENT_RECEIVER_DATA_F;
            break;
        case 3:
            return_value = COMPONENT_RECEIVER_DATA_A;
            break;
        case 4:
            _rpnf->execute(FUNC_TOGGLE_DMOD);
            _lcd->updateStatusDMODE(_rpnf->rpnStack->getDModeName());
            return NO_CHANGE;

        // Column 1 does nothing
        // Column 2 does nothing
        // Column 3 does nothing
        // Column 4 does nothing
        // Column 5 does nothing
        // Column 6 does nothing
        // Column 7
        case 32:
            // erase program memory
            _lcd->outputTerminalLine( 10, "ERASE memory? (\030)");
            if( _getConfirmation( 31)) _rpnf->extMem->clear();

        default: // all other buttons do nothing - keep K-mode
           return NO_CHANGE;
    }
    _mode = 0;
    return return_value;
}

File: MK-52_Simulator/Receivers/Receiver_DATA_K.cs
//////////////////////////////////////////////////////////
//
//  MK-52 RESURRECT
//  Copyright (c) 2020 Mike Yakimov.  All rights reserved.
//  See main file for the license
//
//////////////////////////////////////////////////////////

using System;
using System.Collections.Generic;
using System.Text;
using System.IO;

namespace MK52Simulator
{
    public class Receiver_DATA_K: Receiver_DATA
    {
        public Receiver_DATA_K(MK52_Host parent)
            : base( parent)
        {
            Moniker = "DATA_K";
        }

        public override void activate(string prevReceiver)
        {
            base.activate(prevReceiver);
            LCD_Manager lm = _parent.getLCD();
            lm.updateStatusFMODE(" K ");
            lm.forcePaint();
        }

        public override byte tick(byte scancode)
        {
            RPN_Functions _rpnf = _parent.getFunctions();
            switch (scancode)
            {
                case 0: // keyboard inactive
                    return 0;

                // Column 0
                case 1:
                    _rpnf.requestNextReceiver("DATA_F");
                    return 0;
                case 3:
                    _rpnf.requestNextReceiver("DATA_A");
                    return 0;
                case 4:
                    //    _rpnf.execute(RPN_Functions.FUNC_TOGGLE_DMOD, "");
                    //    base.tick(0);
                    return 0;

                // Column 1 does nothing (for now)
                // Column 2 does nothing (for now)
                // Column 3 does nothing (for now)
                // Column 4 does nothing (for now)
                // Column 5 does nothing (for now)
                // Column 6 does nothing (for now)

                // Column 7
                case 32:
                    _rpnf.requestNextReceiver("Data_Erase");
                    return 0;

                case 33:
                    // Shutdown signal
                    _rpnf.requestNextReceiver("DATA_N");
                    return 33;

                default: // all other buttons do nothing, keeping K-mode
                    return 0;
            }
            //_rpnf.requestNextReceiver("DATA_N");
            //return 0;
        }
    }
}
